stingray/simulator/base.py

```diff
@@ -18,87 +28,152 @@ def simulate_times(lc, use_spline=False, bin_time=None):
     ----------------
     use_spline : bool
         Approximate the light curve with a spline to avoid binning effects
-    bin_time : float
-        The bin time of the light curve, if it needs to be specified for
-        improved precision
 
     Returns
     -------
     times : array-like
         Simulated photon arrival times
-    """
-
-    times = lc.time
-    counts = lc.counts
-
-    bin_time = assign_value_if_none(bin_time, lc.dt)
-    n_bin = len(counts)
-    bin_start = 0
-    maxlc = np.max(counts)
-    intlc = maxlc * n_bin
-    n_events_predict = int(intlc + 10 * np.sqrt(intlc))
-
-    # Max number of events per chunk must be < 100000
-    events_per_bin_predict = n_events_predict / n_bin
-
-    if use_spline:
-        max_bin = int(np.max([4, 1000000 / events_per_bin_predict]))
-
-    else:
-        max_bin = int(np.max([4, 5000000 / events_per_bin_predict]))
-
-    ev_list = np.zeros(n_events_predict)
-    nev = 0
-
-    while bin_start < n_bin:
 
-        t0 = times[bin_start]
-        bin_stop = np.min([bin_start + max_bin, n_bin + 1])
-
-        lc_filt = counts[bin_start:bin_stop]
-        t_filt = times[bin_start:bin_stop]
-        length = t_filt[-1] - t_filt[0]
-
-        n_bin_filt = len(lc_filt)
-        n_to_simulate = n_bin_filt * np.max(lc_filt)
-        safety_factor = 10
+    Examples
+    --------
+    >>> t = [0.5, 1.5, 3.5]
+    >>> c = [100] * 3
+    >>> lc = Lightcurve(t, c, gti=[[0, 2], [3, 4]], dt=1.)
+    >>> times = simulate_times(lc, use_spline=True)
+    >>> np.all(np.diff(times) > 0)  # Output array is sorted
+    True
+    >>> np.all(times >= 0.)  # All times inside GTIs
+    True
+    >>> np.all(times <= 4.)
+    True
+    >>> np.any(times > 3.)
+    True
+    >>> np.any(times < 2.)
+    True
+    >>> np.any((times > 2.) & (times < 3.))  # No times outside GTIs
+    False
+    >>> lc.counts[0] = -3.
+    >>> simulate_times(lc)  # Test with one negative value in the lc
```

Comment:
So what's the commonly accepted solution to generating light curves with values below zero in them? Just set them to zero? Do we have recommendations for that case? If so, we should document them somewhere, because that happens reasonably often. :)

Reply:
They cannot be Poissonian, and they will fail if user does not specify `err_dist=gauss` or similar. But here, we want a light curve that has a number of counts per bin (not necessarily Poisson-distributed, but it has to give a positive number). That is why I let the code throw an exception if people have negative values here.